### PR TITLE
Filter support in Data View API

### PR DESCRIPTION
**Pull request by mshabarov**
These changes in Grid tied with this Data View API updates in Flow https://github.com/vaadin/flow/pull/8874

### CODE REVIEW

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractItemCountGridPage.java [hunk 1]
```diff
@@ -218,7 +218,15 @@ protected void switchToDataProvider() {
     }
 
     protected void switchToDefinedSize() {
-        grid.getLazyDataView().setItemCountCallback(dataProvider::size);
+        grid.getLazyDataView().setItemCountCallback(query -> {
```
Comment (pleku) [inline vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractItemCountGridPage.java:221]: Another option could have been to change the data provider filter to be `Object` instead of `Void` ?

Now I realize the hindrance, luckily the only one, of not having the filter type parameter in `LazyDataView`. It makes this `itemCountCallback` method a bit useless for both `Grid` and for `ComboBox`. 

Since keeping it makes sense for switching between defined/undefined size without a full reset, I wonder if we should just implement it in the concrete LazyDataView classes for the components, so that the type could go right. It would have some code duplication, but then again the type would be immediately the correct one for the users.

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractItemCountGridPage.java [hunk 2]
```diff
@@ -218,7 +218,8 @@ protected void switchToDataProvider() {
     }
 
     protected void switchToDefinedSize() {
-        grid.getLazyDataView().setItemCountCallback(dataProvider::size);
+        grid.getLazyDataView()
+                .setItemCountCallback(query -> dataProvider.size(query));
```
Comment (pleku) [inline vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractItemCountGridPage.java:222]: Use method reference instead, meaning revert unnecessary change

**Review by pleku — APPROVED** [review]

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractItemCountGridPage.java (hunk 2 above)
Comment (mshabarov) [reply vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/dataview/AbstractItemCountGridPage.java]: Hmm, why did I do this...Reverted

**Review by pleku — APPROVED** [review]

**Comment by pleku** [comment]
Now there is a new flow snapshot so the build should be green.

**Comment by mshabarov** [comment]
Components uses `4.0-SNAPSHOT` and master has `4.1-SNAPSHOT`, so let's do the cherry-pick to 4.0
